The Bots!

jisqyv opened the pull request:
The ChatBot and ImageBot combined. Should be merged as two commits (I can do the merging after approval so git signatures remain intact).

ewpatton commented on appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockImageBot.java:
```diff
@@ -0,0 +1,138 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2022-2023 MIT, All rights reserved
```
2023 only?

jisqyv replied:
Done

ewpatton commented on appinventor/appengine/src/com/google/appinventor/client/editor/simple/palette/SimpleComponentDescriptor.java:
```diff
@@ -519,6 +531,8 @@ public static MockComponent createMockComponent(String name, String type, Simple
       return new MockFeatureCollection(editor);
     } else if (name.equals(MockChart.TYPE)) {
       return new MockChart(editor);
+    } else if (name.equals(MockImage.TYPE)) {
+      return new MockChart(editor);
```
????

jisqyv replied:
Yeah ??? Not sure where this came from, but I flushed it. It turns out it is harmless because the check for MockImage.TYPE already happened and if true the function returned. But, yeah. It's gone now.

ewpatton commented on appinventor/docs/markdown/reference/components/experimental.md:
```diff
@@ -8,7 +8,66 @@ title: Experimental
+
+{:.properties}
+
+{:id="ChatBot.ApiKey" .text .ro .bo} *ApiKey*
+: Property for ApiKey
+
+{:id="ChatBot.Apikey" .text .wo .bo} *Apikey*
+: An ApiKey for ChatGPT. User supplied. If provided, we will use it instead of the
+ API key embedded in the chat proxy service.
+
```
Since the Apikey was corrected to ApiKey, rebuilding the documentation with `ant docs` should consolidate these entries.

ewpatton reviewed (changes requested):
A few minor items.

jisqyv commented on appinventor/components/src/com/google/appinventor/components/runtime/ChatBot.java:
```diff
@@ -0,0 +1,580 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2022-2023 MIT, All rights reserved
```
Done

jisqyv commented on appinventor/components/src/com/google/appinventor/components/runtime/ChatBot.java:
```diff
@@ -0,0 +1,580 @@
+  public static final String CHATBOT_SERVICE_URL =
+    "https://chatbot.appinventor.mit.edu/chat/v1";
+
+  private String token;         // MIT Access Token, generated by MockChatBot
```
Done

jisqyv commented on appinventor/components/build.xml:
```diff
@@ -139,6 +139,7 @@
     <copy toFile="${public.deps.dir}/acra-4.4.0.jar" file="${lib.dir}/acra/acra-4.4.0.jar" />
     <copy toFile="${public.deps.dir}/twitter4j.jar" file="${lib.dir}/twitter/twitter4j-core-3.0.5.jar" />
     <copy toFile="${public.deps.dir}/twitter4jmedia.jar" file="${lib.dir}/twitter/twitter4j-media-support-3.0.5.jar" />
+    <copy toFile="${public.deps.dir}/httpcore.jar" file="${lib.dir}/apache-http/httpcore-4.3.2.jar" />
```
Done

jisqyv commented:
@ewpatton I've addressed the review comments. I also changed the comment on ChatBot.ErrorOccurred to make more sense (it said something about the ApiKey being invalid, but given that we don't require one...). I also reference https://appinv.us/chatbot. This doesn't exist yet, but will point to a page on our website with current information (quotas, allowlist information etc.) which might change over time.

jisqyv commented:
@ewpatton If you approve the change, I'll merge/rebase as to protect the digital signatures.

jisqyv commented:
@halatmit I'm not sure, but that isn't a function of the ChatBot as much as it is the speech to text component.
